File: BAEKJOON/2631 - 줄세우기/줄세우기.py
import bisect  # bisect 모듈을 불러옵니다. 이진 탐색을 활용해 LIS를 계산할 때 사용됩니다.

# 입력 받기
n = int(input())  # 첫 번째 줄에서 아이들의 수 N을 입력받습니다.
children = [
    int(input()) for _ in range(n)
]  # 두 번째 줄부터 N개의 숫자를 리스트에 저장합니다.

# LIS 계산
lis = []  # LIS를 저장할 리스트를 초기화합니다.
for num in children:  # 모든 아이들의 번호를 순회합니다.
    pos = bisect.bisect_left(
        lis, num
    )  # 현재 번호가 LIS 리스트에 삽입될 위치를 찾습니다.
    if pos == len(lis):  # 위치가 LIS 리스트의 끝이라면
        lis.append(num)  # 현재 번호를 LIS 리스트에 추가합니다.
    else:  # 그렇지 않다면
        lis[pos] = num  # 기존 값을 현재 번호로 업데이트하여 LIS 리스트를 유지합니다.

# 움직여야 하는 최소 아이들의 수 계산
min_moves = n - len(
    lis
)  # 전체 아이들의 수에서 LIS 길이를 빼서 이동해야 하는 최소 아이들의 수를 계산합니다.
print(min_moves)  # 계산된 최소 이동 수를 출력합니다.

File: BAEKJOON/2631 - 줄세우기/줄세우기.cs
using System;
using System.Collections.Generic;

class Program
{
    static void Main()
    {
        int n = int.Parse(Console.ReadLine()); // 아이들의 수를 입력받습니다.
        int[] children = new int[n]; // 아이들의 번호를 저장할 배열을 선언합니다.

        for (int i = 0; i < n; i++)
        {
            children[i] = int.Parse(Console.ReadLine()); // 아이들의 번호를 입력받아 배열에 저장합니다.
        }

        List<int> lis = new List<int>(); // 최장 증가 부분 수열(LIS)을 저장할 리스트를 선언합니다.

        foreach (int num in children)
        {
            // 현재 숫자가 LIS에 들어갈 위치를 이진 탐색으로 찾습니다.
            int pos = lis.BinarySearch(num);
            if (pos < 0) pos = ~pos; // BinarySearch는 음수를 반환하면 해당 위치를 계산하기 위해 ~를 사용합니다.

            if (pos == lis.Count)
            {
                // 위치가 LIS의 끝이라면 현재 숫자를 추가합니다.
                lis.Add(num);
            }
            else
            {
                // 그렇지 않으면 해당 위치 값을 현재 숫자로 대체합니다.
                lis[pos] = num;
            }
        }

        // 최소 이동 수는 전체 아이들 수에서 LIS 길이를 뺀 값입니다.
        Console.WriteLine(n - lis.Count);
    }
}
